### PR TITLE
Add design for the forking bank

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Each block is a voting point. Each vote can be rolled back.  The Bank needs a way to manage the rollback points. 

#### Summary of Changes

Architecture doc for the fork checkpoint design.

Fixes # (pr #2640)

### CODE REVIEW

File: book/src/fork-checkpoints.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+For example:
+```
+              1
+            /   \
+           2    / \
+          / \  6   7
+         4   5        
+```
+
+The following *acitve chains* are in the checkpoints DAG
```
Comment (garious) [inline book/src/fork-checkpoints.md]: active

File: book/src/fork-checkpoints.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
+# Fork Checkpoints
+
+This design describes the software architecture for fork check pointing.  It addresses the following challenges:
+
+* Forks are created at every block.
+* Forks can be based on any previously produced block.
+* Forks are eventually fully committed such that rollback is impossible.
```
Comment (garious) [inline book/src/fork-checkpoints.md]: "fully committed" -> "finalized"?

File: book/src/fork-checkpoints.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+* Unreachable forks need to be pruned.
+
+## Architecture
+
+The basic design idea is to maintain a DAG of forks.  Each fork points back to a single ancestor.  The DAG is initialized with a root.  Each subsequent fork must descend from the root.  
+
+## Active Chains
+
+An *active chain* is a direct list of connected forks that descend from the current root to a specific fork without any descendants.
```
Comment (garious) [inline book/src/fork-checkpoints.md]: Can this acknowledge the `Blocktree` terminology?  Looks like "active chain" means the same thing as a path from a Blocktree leaf to the root.  Also, why introduce the term "chain" here? Why not an "active fork"?  Use of the word "active" is nicely consistent with GitHub's notion of a "active branch".

File: book/src/fork-checkpoints.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# Fork Checkpoints
+
+This design describes the software architecture for fork check pointing.  It addresses the following challenges:
```
Comment (garious) [inline book/src/fork-checkpoints.md]: checkpointing

File: book/src/fork-checkpoints.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+* Forks are created at every block.
```
Comment (garious) [inline book/src/fork-checkpoints.md]: Are they?

File: book/src/fork-checkpoints.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+An *active chain* is a direct list of connected forks that descend from the current root to a specific fork without any descendants.
+
+For example:
+```
+              1
+            /   \
+           2    / \
+          / \  6   7
+         4   5
```
Comment (rob-solana) [inline book/src/fork-checkpoints.md]: should we stretch this out to reflect the lengths of the required ticks?

```
      1
     / \
    2   \
   /|   |
  / |   |
 4  |   |
    5   /\  
       6  \ 
           7
```

File: book/src/fork-checkpoints.md (hunk 5 above)
Comment (aeyakovenko) [reply book/src/fork-checkpoints.md]: @garious   Every block is a voting point.  Every vote can be rolled back.  A new fork is created at every block.

Comment (garious) [reply book/src/fork-checkpoints.md]: Something is only called a fork if there are two paths.  Maybe you're just missing the word "potentially".

**Comment by garious** [comment]
The problem description for this PR doesn't describe a problem.

**Comment by garious** [comment]
@rob-solana, can you give this another pass? I get the feeling I'm confusing forks and deltas.

**Pull request merged**